replica/rebase.go

```diff
@@ -0,0 +1,245 @@
+package replica
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/renproject/hyperdrive/block"
+	"github.com/renproject/hyperdrive/id"
+	"github.com/renproject/hyperdrive/process"
+)
+
+// BlockStorage extends the `process.Blockchain` interface with the
+// functionality to load the last committed `block.Standard`, and the last
+// committed `block.Base`.
+type BlockStorage interface {
+	process.Blockchain
+
+	LatestBlock() block.Block
+	LatestBaseBlock() block.Block
+}
+
+type BlockDataIterator interface {
+	NextBlockData(block.Kind, Shard) block.Data
+}
+
+type Validator interface {
+	IsBlockValid(block.Block, Shard) bool
+}
+
+type Observer interface {
+	DidCommitBlock(block.Height, Shard)
+}
+
+type shardRebaser struct {
+	mu *sync.Mutex
+
+	expectedKind       block.Kind
+	expectedRebaseSigs id.Signatories
+
+	blockStorage      BlockStorage
+	blockDataIterator BlockDataIterator
+	validator         Validator
+	observer          Observer
+	shard             Shard
+}
+
+func newShardRebaser(blockStorage BlockStorage, blockDataIterator BlockDataIterator, validator Validator, observer Observer, shard Shard) *shardRebaser {
+	return &shardRebaser{
```

Comment:
The `shard` field is not initialised (the function argument is never used).

Reply:
Fixed, thanks!